shell/platform/android/io/flutter/embedding/FlutterActivity.java

```diff
@@ -0,0 +1,238 @@
+ * Android app.
+ * TODO(mattcarroll): how exactly does FlutterActivity need to be leveraged to display Flutter content?
+ *
+ * If Flutter is needed in a location that cannot use an {@code Activity>}, consider using
+ * a {@link FlutterFragment}. Using a {@link FlutterFragment} requires forwarding some calls from
+ * an {@code Activity} to the {@link FlutterFragment}.
+ *
+ * If Flutter is needed in a location that can only use a {@code View}, consider using a
+ * {@link io.flutter.view.FlutterView}. Using a {@link io.flutter.view.FlutterView} requires
+ * forwarding some calls from an {@code Activity}, as well as forwarding lifecycle calls from
+ * an {@code Activity} or a {@code Fragment}.
+ *
+ * @see FlutterFragment
+ *
+ * @see io.flutter.view.FlutterView
+ */
+@SuppressLint("Registered")
+@TargetApi(Build.VERSION_CODES.ICE_CREAM_SANDWICH_MR1)
+public class FlutterActivity extends Activity {
+  private static final String TAG = "FlutterActivity";
+  private static final String TAG_FLUTTER_FRAGMENT = "flutter_fragment";
+  private static final int CONTAINER_ID = 609893468; // random number
+  private static final String SPLASH_SCREEN_META_DATA_KEY = "io.flutter.app.android.SplashScreenUntilFirstFrame";
+
+  private FlutterFragment flutterFragment;
+
+  @Override
+  public void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+
+    FlutterShellArgs args = FlutterShellArgs.fromIntent(getIntent());
+    // TODO(mattcarroll): Change FlutterMain to accept FlutterShellArgs and move additional constants in
+    //       FlutterMain over to FlutterShellArgs.
+    FlutterMain.ensureInitializationComplete(getApplicationContext(), args.toArray());
+
+    configureStatusBarColor();
+    createFlutterFragmentContainer();
+    createFlutterFragment();
+  }
+
+  @Override
+  public void onPostResume() {
+    super.onPostResume();
+    flutterFragment.onPostResume();
+  }
+
+  @Override
+  public void onBackPressed() {
+    flutterFragment.onBackPressed();
+  }
+
+  public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
+    flutterFragment.onRequestPermissionsResult(requestCode, permissions, grantResults);
+  }
+
+  @Override
+  protected void onNewIntent(Intent intent) {
+    boolean isRunAction = Intent.ACTION_RUN.equals(intent.getAction());
+    if (isDebuggable() && isRunAction) {
+      // We're in debug mode, and processing an ACTION_RUN, so we will hot reload Flutter.
+      String route = intent.getStringExtra("route");
+      String appBundlePath = intent.getDataString();
+      if (appBundlePath == null) {
+        // Fall back to the installation path if no bundle path was specified.
+        appBundlePath = FlutterMain.findAppBundlePath(getApplicationContext());
+      }
+      // TODO: What does it mean to have a null appBundlePath?  Should we throw exception in FlutterMain instead of returning null?
```

Comment:
If `FlutterMain.findAppBundlePath` returns null, then something is wrong with how the app bundle assets were packaged into the APK.  Throwing an exception would be reasonable.

Reply:
Thanks. I changed to throw exception in FlutterMain.